cmd/dynamic_startup.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/onflow/flow-go/state/protocol"
+	utilsio "github.com/onflow/flow-go/utils/io"
+
+	"github.com/onflow/flow-go/cmd/util/cmd/common"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol/inmem"
+)
+
+// GetProtocolSnapshot callback that will get latest finalized protocol snapshot
+type GetProtocolSnapshot func() (protocol.Snapshot, error)
+
+// GetSnapshot will attempt to get the latest finalized protocol snapshot with the given flow configs
+func GetSnapshot(ctx context.Context, anAddress, anPub string) (*inmem.Snapshot, error) {
+	// get flow client config that connects to our trusted access node
+	config, err := common.NewFlowClientConfig(anAddress, anPub, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create flow client config for node dynamic startup pre-init: %w", err)
+	}
+
+	flowClient, err := common.FlowClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create flow client for node dynamic startup pre-init: %w", err)
+	}
+
+	b, err := flowClient.GetLatestProtocolStateSnapshot(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest finalized protocol state snapshot during pre-initialization: %w", err)
+	}
+
+	var snapshotEnc inmem.EncodableSnapshot
+	err = json.Unmarshal(b, &snapshotEnc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal protocol state snapshot: %w", err)
+	}
+
+	snapshot := inmem.SnapshotFromEncodable(snapshotEnc)
+	return snapshot, nil
+}
+
+// GetSnapshotAtEpochAndPhase will get the latest finalized protocol snapshot and check the current epoch and epoch phase.
+// If we are past the target epoch and epoch phase we exit the retry mechanism immediately.
+// If not check the snapshot at the specified interval until we reach the target epoch and phase.
+func GetSnapshotAtEpochAndPhase(logger zerolog.Logger, startupEpoch int, startupEpochPhase flow.EpochPhase, retryInterval time.Duration, getSnapshot GetProtocolSnapshot) (protocol.Snapshot, error) {
+	start := time.Now()
+	var snapshot protocol.Snapshot
+
+	constRetry, err := retry.NewConstant(retryInterval)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create retry mechanism: %w", err)
+	}
+
+	err = retry.Do(context.Background(), constRetry, func(ctx context.Context) error {
+		snapshot, err = getSnapshot()
+		if err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to get protocol snapshot: %w", err))
+		}
+
+		// if we encounter any errors interpreting the snapshot something went wrong stop retrying
+		currEpochCounter, err := snapshot.Epochs().Current().Counter()
+		if err != nil {
+			return fmt.Errorf("failed to get the current epoch counter: %w", err)
+		}
+
+		currEpochPhase, err := snapshot.Phase()
+		if err != nil {
+			return fmt.Errorf("failed to get the current epoch phase: %w", err)
+		}
+
+		// check if we are in or past the target epoch and phase
+		if currEpochCounter >= uint64(startupEpoch) && currEpochPhase >= startupEpochPhase {
```

Comment:
If we configure the node to startup at `counter=5,phase=setup`, and we observe a snapshot with `counter=6,phase=staking`, this will evaluate to false. But we would like to start up in this case, because we have passed the target startup time. (Sorry, I oversimplified this condition in a previous comment.)

We would need to change the condition here to:
```
currEpochCounter > startupEpoch || (currEpochCounter == startupEpoch && currEpochPhase >= startupEpochPhase)
```

But, changing this condition has a side effect on the desired property that operators taking the standard path should need to change configuration as little as possible (ideally not at all). These operators joining at epoch N will stake in the staking phase of epoch N-1, then start their node anytime afterward (typically during the staking phase of epoch N-1). With the above change, these operators would need to manually specify their startup epoch (since the default is 0). 

To address this, I suggest we change the default value for `startupEpoch` from `0` to something differentiable from a user input (since 0 is also a valid value to specify manually). If `startupEpoch` is unspecified, we will set it to the current epoch value of the first snapshot we observe. That gives the desired behaviour for the standard usage without needing to manually set any extra flags (to join the network at epoch N, we begin waiting in epoch N-1, and we would like to start up in the setup phase of epoch N-1).

For example, we could add a sentinel value for the flag like `--dynamic-startup-epoch-counter=current` which would have this behaviour of using the epoch counter for the first snapshot we see, and set this to be the default.

Reply:
This `--dynamic-startup-epoch-counter=current` would be nice but there is no way to support this without adding an addtional string var flag, or turning the `--dynamic-startup-epoch-counter` flag into a string and doing some uint64 parsing which is not needed. I like the idea of using 0 to mean "current" epoch . Update the flag usage to reflect this 

`use 0 to default to starting node in the current epoch`

Reply:
https://github.com/onflow/flow-go/pull/1893/commits/62ca33e40cb6e5a2182429d6f647dd7eaca0562d

Reply:
That is a good idea. I'm just worried about one case which will not have the expected behaviour:

I want my node to start up at epoch 0, setup phase. I configure my node with `--dynamic-startup-epoch-counter=0 --dynamic-startup-epoch-phase=EpochPhaseSetup`. 

When I start the process, the network is at epoch 1, staking phase. Because my configuration of the startup epoch counter being set to 0 is interpreted as "current" epoch rather than "epoch with counter 0", my node now won't start up until epoch 1, setup phase (it should start immediately).

This wouldn't be a practical problem on Mainnet, since we're well past epoch 0 and we won't be going back. So maybe it is OK. It would complicate testing a bit though. 

Another thing we could do which would address this edge case is to incorporate an identity table check into the snapshot polling. So, if the default configs were used (epoch=0, phase=setup), and we observe a snapshot where our node ID appears, then also consider that as sufficient reason to start up.

Reply:
With the current implementation your node should start immediately if you 
1. specify an epoch in the past 
2. if you don't specify anything (defaults to 0) 

So during tests if the network is at epoch 1 staking phase and you configure your node to start up with epoch 0 setup phase, your node should start immediately because that epoch has past.